bootstrap/src/main/java/jenkins/bootstrap/Bootstrap.java

```diff
@@ -0,0 +1,313 @@
+                if(value!=null && value.trim().length()>0)
+                    return new FileAndDescription(new File(value.trim()),"JNDI/java:comp/env/"+name);
+                // look at one more place. See issue #1314
+                value = (String) iniCtxt.lookup(name);
+                if(value!=null && value.trim().length()>0)
+                    return new FileAndDescription(new File(value.trim()),"JNDI/"+name);
+            } catch (NamingException e) {
+                // ignore
+            }
+        }
+
+        // next the system property
+        for (String name : HOME_NAMES) {
+            String sysProp = getString(name);
+            if(sysProp!=null)
+                return new FileAndDescription(new File(sysProp.trim()),"SystemProperties.getProperty(\""+name+"\")");
+        }
+
+        // look at the env var next
+        for (String name : HOME_NAMES) {
+            String env = System.getenv(name);
+            if(env!=null)
+                return new FileAndDescription(new File(env.trim()).getAbsoluteFile(),"EnvVars.masterEnvVars.get(\""+name+"\")");
+        }
+
+        // otherwise pick a place by ourselves
+
+        String root = event.getServletContext().getRealPath("/WEB-INF/workspace");
+        if(root!=null) {
+            File ws = new File(root.trim());
+            if(ws.exists())
+                // Hudson <1.42 used to prefer this before ~/.hudson, so
+                // check the existence and if it's there, use it.
+                // otherwise if this is a new installation, prefer ~/.hudson
+                return new FileAndDescription(ws,"getServletContext().getRealPath(\"/WEB-INF/workspace\")");
+        }
+
+        File legacyHome = new File(new File(System.getProperty("user.home")),".hudson");
+        if (legacyHome.exists()) {
+            return new FileAndDescription(legacyHome,"$user.home/.hudson"); // before rename, this is where it was stored
+        }
+
+        File newHome = new File(new File(System.getProperty("user.home")),".jenkins");
+        return new FileAndDescription(newHome,"$user.home/.jenkins");
+    }
+
+    private String getString(String key) {
+        String value = System.getProperty(key); // keep passing on any exceptions
+        if (value != null) {
+            return value;
+        }
+
+        value = context.getInitParameter(key);
+        if (value != null) {
+            return value;
+        }
+
+        return null;
+    }
+
+    /**
+     * This file captures failed boot attempts.
+     * Every time we try to boot, we add the timestamp to this file,
+     * then when we boot, the file gets deleted.
+     */
+    public static File getBootFailureFile(File home) {
+        return new File(home, "failed-boot-attempts.txt");
+    }
+
+    private static final Logger LOGGER = Logger.getLogger(Bootstrap.class.getName());
+
+    private static final String[] HOME_NAMES = {"JENKINS_HOME","HUDSON_HOME"};
```

Comment:
Should we follow the Oracle codestyle and keep these definitions in the beginning of the file? I do not care much, but we declare this codestyle somewhere